src/net/uri.c

```diff
@@ -0,0 +1,294 @@
+
+	/* TODO: validate path */
+	return path;
+}
+
+static bool
+uri_validate(URIParser *parser)
+{
+	if (parser->scheme == URI_SCHEME_INVALID)
+	{
+		parser->error = "invalid URI scheme";
+		return false;
+	}
+
+	if (parser->host == NULL || strlen(parser->host) == 0)
+	{
+		parser->error = "invalid URI host";
+		return false;
+	}
+
+	if (parser->port < 0 || parser->port > INT16_MAX)
+	{
+		parser->error = "invalid URI port";
+		return false;
+	}
+
+	return true;
+}
+
+static void
+parser_set_state(URIParser *parser, URIParseState new_state)
+{
+	parser->prev_state = parser->state;
+	parser->state = new_state;
+	parser->tail = parser->head + 1;
+}
+
+static bool
+parser_handle_colon(URIParser *parser)
+{
+	switch (parser->state)
+	{
+		case PARSE_STATE_SCHEME:
+			Assert(parser->prev_state == PARSE_STATE_SCHEME);
+			parser->scheme = parse_scheme(parser);
+			parser_set_state(parser, PARSE_STATE_AUTHORITY_SLASH1);
+			break;
+		case PARSE_STATE_HOST:
+			Assert(parser->prev_state == PARSE_STATE_AUTHORITY_SLASH2);
+			parser->host = parse_host(parser);
+			parser_set_state(parser, PARSE_STATE_PORT);
+			break;
+		default:
+			return false;
+	}
+
+	return true;
+}
+
+static bool
+parser_handle_slash(URIParser *parser)
+{
+	switch (parser->state)
+	{
+		case PARSE_STATE_AUTHORITY_SLASH1:
+			Assert(parser->prev_state == PARSE_STATE_SCHEME);
+			parser_set_state(parser, PARSE_STATE_AUTHORITY_SLASH2);
+			break;
+		case PARSE_STATE_AUTHORITY_SLASH2:
+			Assert(parser->prev_state == PARSE_STATE_AUTHORITY_SLASH1);
+			parser_set_state(parser, PARSE_STATE_HOST);
+			break;
+		case PARSE_STATE_HOST:
+			Assert(parser->prev_state == PARSE_STATE_AUTHORITY_SLASH2);
+			parser->host = parse_host(parser);
+			parser_set_state(parser, PARSE_STATE_PATH);
+			break;
+		case PARSE_STATE_PORT:
+			Assert(parser->prev_state == PARSE_STATE_HOST);
+			parser->port = parse_port(parser);
+			parser_set_state(parser, PARSE_STATE_PATH);
+			break;
+		case PARSE_STATE_PATH:
+			/* TODO: check for supported path characters */
+			break;
+		default:
+			return false;
+	}
+	return true;
+}
+
+static bool
+parser_handle_questionmark(URIParser *parser)
+{
+	switch (parser->state)
+	{
+		case PARSE_STATE_HOST:
+		case PARSE_STATE_PORT:
+		case PARSE_STATE_PATH:
+			parser->error = "URI does not support query parameters";
+		default:
+			return false;
+	}
+	return true;
+}
+
+static bool
+parser_handle_end(URIParser *parser)
+{
+	switch (parser->state)
+	{
+		case PARSE_STATE_PORT:
+			Assert(parser->prev_state == PARSE_STATE_HOST);
+			parser->port = parse_port(parser);
+			break;
+		case PARSE_STATE_HOST:
+			Assert(parser->prev_state == PARSE_STATE_AUTHORITY_SLASH2);
+			parser->host = parse_host(parser);
+			break;
+		case PARSE_STATE_PATH:
+			Assert(parser->prev_state == PARSE_STATE_HOST);
+			parser->path = parse_path(parser);
+			break;
+		default:
+			return false;
+	}
+
+	return true;
+}
+
+URI *
+uri_parse(const char *raw_uri, const char **errhint)
```

Comment:
This seems a bit over-engineered for what is (note completely untested code) essentially:

```C
    /* input: char *raw */
    char *hostname_start = strstr(raw, "://");
    size_t hostname_end = strcspn(raw, ":/");
    long port = 0;
    char *path = NULL;
    switch raw[hostname_end] {
        case ':':
            port = strtol(raw + hostname_end + 1, &path, 10);
            break;
        case '/':
            path = raw + hostname_end + 1;
        default:
            break;
    }
    /* copy hostname into its own string, and create the struct */
```

Is there a reason we need the entire URI state-machine and validation here instead of letting `getaddrinfo` do it? The latter would be strictly more forwards compatible if we support additional URIs, and much easier to maintain.

Reply:
AFIAK, `getaddrinfo` does not parse URIs. It's used for name resolution, basically resolving a hostname to an IP address and looking at `/etc/services` for service mappings. Actually not really sure what you are proposing, since the URI parser doesn't do any name resolution like `getaddrinfo`. Instead, it's basically giving you something you could pass into `getaddrinfo`.

Reply:
I'm trying to say that: I don't think we need or particularly want to validate the URI here, such validations are better handled at higher layers. For `getaddrinfo` we only need the `host` and `port`: if we give `getaddrinfo` and invalid hostname or scheme it'll treat it just like any other invalid host lookup, and return an error that it either cannot find that host, or support that scheme; our connection logic is the only thing that needs to know what kinds of services we use, it needs to know about all of them since it needs to be able to connect to both `HTTP`, `HTTPS`, and if we want to support a new protocol that layer will have to be changed regardless of what we do here. This being the case, I think it's simpler, and more forward compatible, for the URI logic to _only_ extract something that looks like `(service, host, port, path)`, ignoring whether the input even approaches a valid URI, since higher layers will catch that. The logic for this (which i attempted in C in my earlier comment) is essentially
```
Look for a '://':
  if found: everything before that is the service
  else: error or default to https

In what remains look for a ':' or '\';
  case ':': parse whats next as a port
  case '\' everything remaining is the path; break;
```
which seems to result in much simpler and shorter code than what we currently have.

Reply:
I don't think there is any legit reason to even have this logic, we should #define the url and be done. Why is there any reason to allow the user to configure this?